File: backendFastAPI/app/core/logger.py
"""
Logger propio con redacción de secretos — equivalente de backend/src/utils/logger.js.

Sin dependencias externas, igual que en Node. Silencioso en ENVIRONMENT=test.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from app.core.config import settings

_COLORS = {"debug": "\x1b[90m", "info": "\x1b[36m", "warn": "\x1b[33m", "error": "\x1b[31m"}
_RESET = "\x1b[0m"

_SENSITIVE_KEYS = {
    "password",
    "passwordHash",
    "password_hash",
    "token",
    "accessToken",
    "refreshToken",
    "authorization",
}


def _redact(value: Any) -> Any:
    if isinstance(value, list):
        return [_redact(item) for item in value]
    if isinstance(value, dict):
        return {
            key: ("[REDACTED]" if key in _SENSITIVE_KEYS else _redact(val))
            for key, val in value.items()
        }
    return value


def _log(level: str, message: str, meta: Any = None) -> None:
    if settings.is_test:
        return
    timestamp = datetime.now(timezone.utc).isoformat()
    color = _COLORS.get(level, "")
    prefix = f"{color}[{timestamp}] {level.upper()}{_RESET}"
    if meta is not None:
        print(prefix, message, _redact(meta))
    else:
        print(prefix, message)


class Logger:
    def debug(self, message: str, meta: Any = None) -> None:
        _log("debug", message, meta)

    def info(self, message: str, meta: Any = None) -> None:
        _log("info", message, meta)

    def warn(self, message: str, meta: Any = None) -> None:
        _log("warn", message, meta)

    def error(self, message: str, meta: Any = None) -> None:
        _log("error", message, meta)


logger = Logger()

File: backend/src/utils/logger.js
import { isTest } from '../config/env.js';

const LEVELS = { debug: 0, info: 1, warn: 2, error: 3 };
const COLORS = { debug: '\x1b[90m', info: '\x1b[36m', warn: '\x1b[33m', error: '\x1b[31m' };
const RESET = '\x1b[0m';

const SENSITIVE_KEYS = new Set([
    'password',
    'passwordHash',
    'password_hash',
    'token',
    'accessToken',
    'refreshToken',
    'authorization',
]);

/** Reemplaza recursivamente los valores de claves sensibles antes de imprimir. */
function redact(value) {
    if (Array.isArray(value)) return value.map(redact);
    if (value && typeof value === 'object') {
        const clone = {};
        for (const [key, val] of Object.entries(value)) {
            clone[key] = SENSITIVE_KEYS.has(key) ? '[REDACTED]' : redact(val);
        }
        return clone;
    }
    return value;
}

function log(level, message, meta) {
    if (isTest) return;
    const timestamp = new Date().toISOString();
    const color = COLORS[level] ?? '';
    const prefix = `${color}[${timestamp}] ${level.toUpperCase()}${RESET}`;

    if (meta !== undefined) {
        console.log(prefix, message, redact(meta));
    } else {
        console.log(prefix, message);
    }
}

export const logger = {
    debug: (message, meta) => log('debug', message, meta),
    info: (message, meta) => log('info', message, meta),
    warn: (message, meta) => log('warn', message, meta),
    error: (message, meta) => log('error', message, meta),
};

export default logger;
export { LEVELS };
